.github/workflows/test.yml

```
@@ -77,6 +80,35 @@ jobs:
           files: service-${{ matrix.service }}/build/*.xml
           test_file_prefix: +service-${{ matrix.service }}/
           check_name: ${{ matrix.service }} ${{ matrix.suite }} tests
+      - name: archive test results
+        uses: actions/upload-artifact@v4.6.0
+        with:
+          name: service-${{ matrix.service }}
+          path: service-${{ matrix.service }}/build
+        if: matrix.suite == 'unit-test'
+  code_coverage:
+    strategy:
+      matrix:
+        service: [api, front]
+    runs-on: ubuntu-latest
+    needs: test
+    steps:
+      - uses: actions/checkout@85e6279cec87321a52edac9c87bce653a07cf6c2 # pin@v3
```

Better to use named versions so Renovate can auto-update
```suggestion
      - uses: actions/checkout@v4
```